normandy/base/middleware.py

```diff
@@ -128,3 +136,122 @@ def process_request(self, request):
+            r'Bearer\s+(.*)',
+            request.META.get('HTTP_AUTHORIZATION') or '',
+            re.I
+        )
+        if not matched:
+            # That's cool. If it's not there, it's not there.
+            # This middleware will not be able to authenticate the user
+            # for the duration of this request.
+            return
+        access_token = matched.group(1)
+        cache_key = 'oidc-access-token-check:{}'.format(
+            hashlib.md5(force_bytes(access_token)).hexdigest()
+        )
+        checked = cache.get(cache_key)
+        if checked:
+            return checked
+        # print("access_token", repr(access_token))
+        # XXX This should use a retrying requests sesssion or backoff lib.
+        response = requests.get(self.oidc_user_endpoint, headers={
+            'Authorization': 'Bearer {0}'.format(access_token)
+        })
+        if response.status_code == 200:
+            # Awesome! Turn this into a Django User
+            user_profile = response.json()
+            if 'email' not in user_profile:
+                raise ImproperlyConfigured(
+                    "The HTTP Authorization header had an access token that "
+                    "was created without the 'email' scope."
+                )
+            email = user_profile['email']
+            if not email:
+                # This would happen if someone has requested an access token
+                # from their OIDC provider *without the 'email' scope*.
+                # For example, if you do:
+                #
+                # new auth0.WebAuth({
+                #   domain: OIDC_DOMAIN,
+                #   clientID: OIDC_CLIENT_ID,
+                #   responseType: 'token id_token',
+                #   scope: 'openid profile',       <-- note lack of 'email'
+                # });
+                #
+                # Ideally the client should review their config. The
+                # access token did work but not sufficient for our needs.
+                raise PermissionDenied("User profile lacks 'email' scope.")
+            try:
+                user = self.UserModel.objects.get(email__iexact=email)
+            except self.UserModel.DoesNotExist:
+                user = self.UserModel.objects.create(
+                    username=default_username(email),
+                    email=email,
+                )
+            except self.UserModel.MultipleObjectsReturned:
+                raise PermissionDenied('The email is not unique')
+            if (
+                'family_name' in user_profile and 'given_name' in user_profile and
+                (
+                    user_profile['family_name'] != user.last_name or
+                    user_profile['given_name'] != user.first_name
+                )
+            ):
+                user.first_name = user_profile['given_name']
+                user.last_name = user_profile['family_name']
+                user.save()
```

Comment:
updating the first/last name of the user in the database introduces a side effect.  Seems weird to have it here.